plugins/copr.py

```diff
@@ -502,22 +506,29 @@ def _runtime_deps_warning(self, copr_username, copr_projectname):
         if not runtime_deps:
             return
 
-        print(_("Maintainer of the enabled Copr repository decided to make\n"
-                "it dependent on other repositories. Such repositories are\n"
-                "usually necessary for successful installation of RPMs from\n"
-                "the main Copr repository (runtime dependencies provided).\n\n"
-
-                "Be aware that the note about quality and bug-reporting\n"
-                "above applies here too, Fedora Project doesn't control the\n"
-                "content. Please review the list:"))
-        dep_idx = 1
-        for repo_id in runtime_deps:
-            print("\n{0}. [{1}]".format(dep_idx, repo_id))
-            print("   baseurl={0}".format(repo.cfg.getValue(repo_id, "baseurl")))
-            dep_idx += 1
-
-        msg = _("\nThese repositories have been enabled automatically."
-                "\nDo you want to keep them enabled?")
+        msg = _(
+            "\nMaintainer of the enabled Copr repository decided to make\n"
+            "it dependent on other repositories. Such repositories are\n"
+            "usually necessary for successful installation of RPMs from\n"
+            "the main Copr repository (runtime dependencies provided).\n\n"
+
+            "Be aware that the note about quality and bug-reporting\n"
+            "above applies here too, Fedora Project doesn't control the\n"
+            "content. Please review the list:\n\n"
+            "{0}\n\n"
+            "These repositories have been enabled automatically.\n\n"
+        )
+
+        counter = itertools.count(1)
+        sys.stderr.write(msg.format("\n\n".join([
+            "{num:2}. [{repoid}]\n    baseurl={baseurl}".format(
+                num=next(counter),
+                repoid=repoid,
+                baseurl=repo.cfg.getValue(repoid, "baseurl"))
+            for repoid in runtime_deps
+        ])))
+
+        msg = _("Do you want to keep them enabled?")
```

Comment:
The issue I find here is that the information above is on `stderr` and the prompt here is on `stdout`, so they can end up being separated and the prompt doesn't make sense without the information.

In case I wasn't clear in my previous comment (https://github.com/rpm-software-management/dnf-plugins-core/pull/388#issuecomment-630022872), I meant to either print everything to `stdout` (interactive mode) or everything to `stderr` (non-interactive). But I'm not too happy about that either, that makes it inconsistent in a different way.

Reply:
> The issue I find here is that the information above is on stderr and the
> prompt here is on stdout, so they can end up being separated and the
> prompt doesn't make sense without the information.

How likely is that you redirect stderr, and still want to work
interactively?  I don't think we want to spent too much time on such
details...  which are kind of precedent I think.  I think we should
make a compromise fix here, and wait till there's proper tooling
in dnf for such warnings.

As you pointed out, some software in the wild might already read the
stdout...  and since we used logger.info before, I think using just stdout
in the #388 was mistake.  I think there are two options, either use logger
(my slight preference, since nobody complaint so far and we were just
fine) or switch to sys.stderr.write.  Please tell me your preference.

Reply:
Meh, `logger.info` goes to stdout, interesting..  Which means that IMO we
should use sys.stderr ... eventually.

Reply:
Still ... if you have preference, I don't see a problem neither way.

Reply:
> ... and wait till there's proper tooling in dnf for such warnings.

Not wanting to derail much here, but what would be the proper tooling? I think we are talking conceptually here, you print to one or the other, but it always seems to be a compromise. At least I'm not sure what a clean solution is.

I've come to the conclusion (at least for the moment, not saying I can't be convinced) that using logger for console output is a bad idea. Things you want to go to the logger and to console often differ, what goes to console you don't want to go to the logger and vice versa. (that said outputting logger lines to stderr with e.g.`--debug` is a fine)

Yes, logger.info going to `stdout` is bad.

I'm also thinking whether we should print the information text to `stderr` and actually list the repositories to `stdout`?

However, if you just want to be done with this, we can keep the current version, i.e. printing it to `stderr` (except for the "[y/N]" prompt).

Reply:
> what would be the proper tooling? 

I mean some dnf api like `self.print_user_message(...)` which will do exactly
what DNF people expect it to do (the point is to abstract this stdout/stderr
problem out so plugin contributors don't have to care).

> I'm also thinking whether we should print the information text to stderr and actually list the repositories to stdout?

I don't know.  Feel free to submit PR if you want to make it even nicer.

> However, if you just want to be done with this, we can keep the current version, i.e. printing it to stderr (except for the "[y/N]" prompt).

That's what I prefer, yes.  I don't think anything else is better :-/

Reply:
> I mean some dnf api like `self.print_user_message(...)` which will do exactly
> what DNF people expect it to do (the point is to abstract this stdout/stderr
> problem out so plugin contributors don't have to care).

Good point.

> > However, if you just want to be done with this, we can keep the current version, i.e. printing it to stderr (except for the "[y/N]" prompt).
> 
> That's what I prefer, yes. I don't think anything else so far discussed is better :-/

Okay.